accord-core/src/main/java/accord/primitives/Timestamp.java

```diff
@@ -19,62 +19,126 @@
 package accord.primitives;
 
 import accord.local.Node.Id;
+import accord.utils.Invariants;
 
 public class Timestamp implements Comparable<Timestamp>
 {
+    public static Timestamp fromBits(long msb, long lsb, Id node)
+    {
+        return new Timestamp(msb, lsb, node);
+    }
+
+    public static Timestamp fromRaw(long epoch, long hlc, Id node)
+    {
+        return new Timestamp(epoch, hlc, 0, node);
+    }
+
+    public static Timestamp fromRaw(long epoch, long hlc, int flags, Id node)
+    {
+        return new Timestamp(epoch, hlc, flags, node);
+    }
+
+    public static Timestamp maxForEpoch(long epoch)
+    {
+        return new Timestamp((epoch << 16) | 0xffff, Long.MAX_VALUE, Id.MAX);
+    }
+
+    public static Timestamp minForEpoch(long epoch)
+    {
+        return new Timestamp(epoch << 16, 0, Id.NONE);
+    }
+
+    private static final long MAX_EPOCH = (1L << 48) - 1;
+    private static final long HLC_INCR = 1L << 16;
+    private static final long MAX_FLAGS = HLC_INCR - 1;
     public static final Timestamp NONE = new Timestamp(0, 0, 0, Id.NONE);
 
-    public final long epoch;
-    public final long real;
-    public final int logical;
+    public final long msb;
+    public final long lsb;
     public final Id node;
 
-    public Timestamp(long epoch, long real, int logical, Id node)
+    Timestamp(long epoch, long hlc, int flags, Id node)
     {
-        this.epoch = epoch;
-        this.real = real;
-        this.logical = logical;
+        Invariants.checkArgument(epoch <= MAX_EPOCH);
+        Invariants.checkArgument(flags <= MAX_FLAGS);
+        this.msb = (epoch << 16) | (hlc >>> 48);
+        this.lsb = (hlc << 16) | flags;
+        this.node = node;
+    }
+
+    Timestamp(long msb, long lsb, Id node)
+    {
+        this.msb = msb;
+        this.lsb = lsb;
         this.node = node;
     }
 
     public Timestamp(Timestamp copy)
     {
-        this.epoch = copy.epoch;
-        this.real = copy.real;
-        this.logical = copy.logical;
+        this.msb = copy.msb;
+        this.lsb = copy.lsb;
+        this.node = copy.node;
+    }
+
+    Timestamp(Timestamp copy, int flags)
+    {
+        Invariants.checkArgument(flags <= MAX_FLAGS);
+        this.msb = copy.msb;
+        this.lsb = (copy.lsb & ~MAX_FLAGS) | flags;
         this.node = copy.node;
     }
 
+    public long epoch()
+    {
+        return msb >>> 16;
+    }
+
+    /**
+     * A hybrid logical clock with implementation-defined resolution
+     */
+    public long hlc()
+    {
+        return (msb << 48) | (lsb >>> 16);
+    }
+
+    public int flags()
+    {
+        return (int) (lsb & MAX_FLAGS);
+    }
+
     public Timestamp withMinEpoch(long minEpoch)
```

Comment:
Maybe rename to `withEpochAtLeast()` or something along those lines? My initial understanding was: with min( epoch) out of {current, provided}, but it does the opposite of course.